service/s3/marshal.go

```diff
@@ -0,0 +1,162 @@
+		encoder.AddHeader("x-amz-object-lock-legal-hold").String(string(input.ObjectLockLegalHoldStatus))
+	}
+	if input.ObjectLockMode != "" {
+		encoder.AddHeader("x-amz-object-lock-mode").String(string(input.ObjectLockMode))
+	}
+	if input.Tagging != nil {
+		encoder.AddHeader("x-amz-tagging").String(*input.Tagging)
+	}
+	if input.RequestPayer != "" {
+		encoder.AddHeader("x-amz-request-payer").String(string(input.RequestPayer))
+	}
+	if input.SSEKMSEncryptionContext != nil {
+		encoder.AddHeader("x-amz-server-side-encryption-context").String(*input.SSEKMSEncryptionContext)
+	}
+	if input.SSEKMSKeyId != nil {
+		encoder.AddHeader("x-amz-server-side-encryption-aws-kms-key-id").String(*input.SSEKMSKeyId)
+	}
+	if input.SSECustomerKey != nil {
+		encoder.AddHeader("x-amz-server-side-encryption-customer-key-MD5").String(*input.SSECustomerKeyMD5)
+	}
+	if input.SSECustomerKeyMD5 != nil {
+		encoder.AddHeader("x-amz-server-side-encryption-customer-key-MD5").String(*input.SSECustomerKeyMD5)
+	}
+	if input.SSECustomerAlgorithm != nil {
+		encoder.AddHeader("x-amz-server-side-encryption-customer-algorithm").String(*input.SSECustomerAlgorithm)
+	}
+	if input.WebsiteRedirectLocation != nil {
+		encoder.AddHeader("x-amz-website-redirect-location").String(*input.WebsiteRedirectLocation)
+	}
+	if input.StorageClass != "" {
+		encoder.AddHeader("x-amz-storage-class").String(string(input.StorageClass))
+	}
+	if input.ServerSideEncryption != "" {
+		encoder.AddHeader("x-amz-server-side-encryption").String(string(input.ServerSideEncryption))
+	}
+	if input.Expires != nil {
+		if err := encoder.AddHeader("Expires").Time(*input.Expires, protocol.RFC822TimeFormatName); err != nil {
+			return awserr.New(aws.ErrCodeSerialization, "failed to marshal header for shape Expires using REST encoder", err)
+		}
+	}
+	if input.ObjectLockRetainUntilDate != nil {
+		if err := encoder.AddHeader("x-amz-object-lock-retain-until-date").Time(*input.ObjectLockRetainUntilDate, protocol.ISO8601TimeFormatName); err != nil {
+			return awserr.New(aws.ErrCodeSerialization, "failed to marshal header for shape ObjectLockRetainUntilDate using REST encoder", err)
+		}
+	}
+	//  Encoding shapes with location `uri`
+	if input.Bucket != nil {
+		if err := encoder.SetURI("Bucket").String(*input.Bucket); err != nil {
+			return awserr.New(aws.ErrCodeSerialization, "failed to marshal URI for shape Bucket using REST encoder", err)
+		}
+	}
+	if input.Key != nil {
+		if err := encoder.SetURI("Key").String(*input.Key); err != nil {
+			return awserr.New(aws.ErrCodeSerialization, "failed to marshal URI for shape Key using REST encoder", err)
+		}
+	}
+	return nil
+}
+
+// marshalPutObjectInputShapeAWSXML is a stand alone function used to marshal the xml payload
+// This should be generated according to the payload type for rest-xml protocol
+func marshalPutObjectInputShapeAWSXML(r *aws.Request, input *PutObjectInput) error {
+	if input.Body != nil {
+		r.SetReaderBody(input.Body)
+	}
+	return nil
+}
```

Comment:
We'll want to consider how  the request body is used across multiple request attempts. This is handled in an awkward way within the SDK's `Request` type, but leads to the SDK requiring the `input.Body` to be an `io.ReadSeeker`. Ideally the SDK should always except an `io.Reader`, and fail if the request requires a hash to be computed of the body, or disable retries if the body cannot be seeked.